File: gtci/01-two-pointers/python3/5-triplet-sum-close-to-target.py
import math


class Solution:
    def searchTriplet(self, arr, target_sum):
        closest_sum = None
        closest_distance = None

        arr.sort()

        for idx, num in enumerate(arr):
            if idx > 0 and num == arr[idx - 1]:
                continue

            left = idx + 1
            right = len(arr) - 1

            while left < right:
                current_sum = num + arr[left] + arr[right]
                current_distance = abs(target_sum - current_sum)

                if current_sum == target_sum:
                    return current_sum

                if closest_sum is None:
                    closest_sum = current_sum
                    closest_distance = current_distance
                elif current_distance < closest_distance or (
                    current_distance <= closest_distance and current_sum < closest_sum
                ):
                    closest_sum = current_sum
                    closest_distance = current_distance

                if current_sum < target_sum:
                    left += 1
                elif current_sum > target_sum:
                    right -= 1

        return closest_sum


if __name__ == "__main__":
    sol = Solution()
    arr = [-2, -1, 0, 1, 2]
    target_sum = 0
    assert sol.searchTriplet(arr, target_sum) == 0

    arr2 = [-3, -1, 1, 2]
    target_sum2 = 1
    assert sol.searchTriplet(arr2, target_sum2) == 0

    arr3 = [1, 0, 1, 1]
    target_sum3 = 100
    assert sol.searchTriplet(arr3, target_sum3) == 3

    print("All test cases passed.")

File: gtci/01-two-pointers/cpp/5-triplet-sum-close-to-target.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

class Solution {
   public:
    int searchTriplet(vector<int>& arr, int targetSum) {
        sort(arr.begin(), arr.end());

        int closestSum = 0;
        int closestDistance = numeric_limits<int>::max();

        for (int i = 0; i < arr.size(); i++) {
            const int firstNum = arr[i];

            vector<int>::const_iterator leftItr = arr.cbegin() + i + 1;
            vector<int>::const_iterator rightItr = arr.cend() - 1;

            while (leftItr < rightItr) {
                const int currentSum = firstNum + *leftItr + *rightItr;
                const int currentDistance = abs(targetSum - currentSum);

                if (currentSum == targetSum) return currentSum;

                if (currentDistance < closestDistance ||
                    (currentDistance <= closestDistance &&
                     currentSum < closestSum)) {
                    closestSum = currentSum;
                    closestDistance = currentDistance;
                }

                if (currentSum < targetSum)
                    leftItr++;
                else if (currentSum > targetSum)
                    rightItr--;
            }
        }
        return closestSum;
    }
};

int main() {
    Solution sol;
    std::vector<int> arr1 = {-2, 0, 1, 2};
    assert(sol.searchTriplet(arr1, 2) == 1);

    std::vector<int> arr2 = {1, 0, 1, 1};
    assert(sol.searchTriplet(arr2, 100) == 3);

    std::vector<int> arr3 = {0, 0, 0};
    assert(sol.searchTriplet(arr3, 1) == 0);

    return 0;
}
